woocommerce/woocommerce-ios
# [Bundle products in order form] Product selector changes

Part of #10428 


## Why


Ref: Pe5pgL-3Vd-p2#product-selector-changes

Previously, it was assumed that all order items have unique product/variation IDs. The product selector's selected products/variations also used to include the ones from a bundle product item's child items. This prevents the merchant from selecting the child item product/variation as a standalone order item. Thus, there are a few changes to the product selector:

- The products/variations from any bundle product item’s child items aren’t included in the selected products/variations
- A bundle product row can’t be deselected – removing a bundle product order item needs to be performed outside of the product selector
- Tapping on a bundle product row brings up the Configure screen, while tapping on other rows toggles the selection state
- It’s possible to select & configure the same bundle product multiple times, and all the configurations are saved when the merchant completes the product selector

## How

First, a `Configure` CTA is added to the product row in `ProductSelectorView` if the row is configurable (`isConfigurable` is `true`). In the view model `ProductRowViewModel`, another condition is added to `isConfigurable` to include the `configure` closure being non-nil. When tapping on a configurable row in the product selector, `configure` is invoked instead of `viewModel.changeSelectionStateForProduct`. `configure` is set to a new optional closure in `ProductSelectorViewModel` for its caller to implement how a product can be configured. `EditableOrderViewModel`, the caller of `ProductSelectorViewModel`, implements the configuration to set a view model `productToConfigureViewModel` for the Configure screen.

When the merchant taps to configure a bundle product in the product selector, `EditableOrderViewModel`'s newly set view model `productToConfigureViewModel` is used to present a Configure sheet over the product selector in the `OrderForm` UI. When the merchant taps to save the bundle configuration, it saves the bundle configuration into a private variable that tracks a dictionary from a product ID to an array of bundle configurations (each bundle product can be configured multiple times). Then, when the merchant taps to confirm the product selection, the configuration is popped from the array to set the bundle configuration of the new order item for syncing. If the merchant decides to discard the changes in the product selector, the dictionary variable is cleared.

There are also a few changes in `EditableOrderViewModel` to check an order item's `parent` so that the bundle's child items aren't considered as selected in the product selector.

## Testing instructions


Prerequisite: the store has [Product Bundles extension](https://woocommerce.com/products/product-bundles/) enabled, and also a bundle product with non-empty items.

- Go to the Orders tab
- Tap `+` to create an order
- Tap `Add Products` and select a bundle product with at least one item --> a Configure screen should be shown
- Configure the bundle if you like, and tap `Done` (validation is in https://github.com/woocommerce/woocommerce-ios/pull/11018) --> the product selector should show `1 Product Selected` at the bottom
- Tap on the same bundle product again --> the bundle product remains checked (selected), and a Configure screen should be shown
- Configure the bundle differently from the first time like a different quantity, and tap `Done` --> the product selector should show `2 Products Selected` at the bottom
- Search and select one of the bundle product's item product --> it should be selected after the tap, and the product selector should show `3 Products Selected` at the bottom
- Tap the `3 Products Selected` at the bottom --> the 2 bundle products and 1 non-bundle product should be added
- Tap on `+` in the products header again to add more products --> the product selector should show `3 Products Selected` at the bottom
- Tap on the same bundle product again --> a Configure screen should be shown
- Configure the bundle if you like, and tap `Done` --> the product selector should show `4 Product Selected` at the bottom
- Tap the `4 Products Selected` at the bottom --> the newly configured bundle product should be appended
- After all order item products are ready, tap `Create` --> the 4 products should be added remotely

## Screenshots



https://github.com/woocommerce/woocommerce-ios/assets/1945542/bf4688b6-cb3b-4a8a-8fa1-f804994c3a42



---
